File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums, limit):
        n = len(nums)
        arr = sorted((x, i) for i, x in enumerate(nums))
        
        groups = []
        current = []
        
        for i in range(n):
            if i == 0 or arr[i][0] - arr[i - 1][0] <= limit:
                current.append(arr[i])
            else:
                groups.append(current)
                current = [arr[i]]
        
        groups.append(current)
        
        ans = [0] * n
        
        for group in groups:
            values = sorted(x for x, _ in group)
            indices = sorted(i for _, i in group)
            
            for i in range(len(group)):
                ans[indices[i]] = values[i]
        
        return ans

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<pair<int, int>> a(n);

        for (int i = 0; i < n; ++i)
            a[i] = {nums[i], i};

        sort(a.begin(), a.end());

        vector<int> ans(n);

        int i = 0;
        while (i < n) {
            int j = i;

            while (j + 1 < n &&
                   (long long)a[j + 1].first - a[j].first <= limit) {
                ++j;
            }

            vector<int> indices;

            for (int k = i; k <= j; ++k)
                indices.push_back(a[k].second);

            sort(indices.begin(), indices.end());

            for (int k = 0; k < (int)indices.size(); ++k)
                ans[indices[k]] = a[i + k].first;

            i = j + 1;
        }

        return ans;
    }
};